[Pratham Jain] iP

pratham31012002 opened the pull request:
# Pluto
> "The key is not to prioritise what's on your schedule, but to schedule your priorities." - Stephen Covey ([source](https://www.forbes.com/sites/kevinkruse/2012/07/16/the-7-habits/?sh=3e0f4d839c6d))

**Pluto** frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/pratham31012002/ip/releases/download/A-Jar/pluto.jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you.

And it is **FREE!**
Features:

- [x] Managing tasks :+1:
- [x] Managing Deadlines :+1:
- [x] Adding reminders for events :+1:

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

pratham31012002 committed:
initial commit

pratham31012002 committed:
feat: added command list

pratham31012002 committed:
feat: created Task

pratham31012002 committed:
feat: added mark unmark

pratham31012002 committed:
feat: added Todo task

pratham31012002 committed:
feat: added event and deadline task

pratham31012002 committed:
feat: completed types of tasks

pratham31012002 committed:
feat: Automated Text UI Testing done

pratham31012002 committed:
fix: started error handling

pratham31012002 committed:
feat: added delete feature

pratham31012002 committed:
Add storage of tasks

pratham31012002 committed:
Add formatting for dates

pratham31012002 committed:
Merge branch 'branch-Level-7'
Add storage of tasks created

pratham31012002 committed:
Merge branch 'branch-Level-8'

pratham31012002 committed:
Use enums

pratham31012002 committed:
OOP everything

pratham31012002 committed:
Add show command

pratham31012002 committed:
Add files to packages

pratham31012002 committed:
Fix access modifiers

Nephelite commented on src/main/java/pluto/Parser.java:
```diff
@@ -0,0 +1,115 @@
+            case BYE:
+                return new ExitCommand();
+            default:
+                throw new PlutoException("\tOOPS!!! I'm sorry, but I don't know what that means :-(");
```
I didn't like this indentation of ```case``` clauses in your ```switch``` statement.

According to the module's coding standards, ```case``` clauses share the same indentation as your ```switch``` statement.

I will only mention this here, but please do change all the switch statements to which this applies, such as the one below this block.

Nephelite commented on src/main/java/pluto/Parser.java:
```diff
@@ -0,0 +1,115 @@
+package pluto;
+
+import pluto.command.*;
```
I didn't like the use of wildcard import. 

Even if you are using all files under ```pluto.command```, perhaps consider explicitly importing everything to comply with module coding conventions?

Nephelite commented on src/main/java/pluto/Parser.java:
```diff
@@ -0,0 +1,115 @@
+        } catch (NumberFormatException e) {
+            throw new PlutoException("\tOOPS!!! Index of task is required.");
+        }
+    }
```
I like that you put everything related to parsing into this Parser class. Good use of Abstraction!

Nephelite commented on src/main/java/pluto/Parser.java:
```diff
@@ -0,0 +1,115 @@
+        BYE
+    }
+
+    public static Command parse(String inputLine) throws PlutoException {
```
I didn't like how undescriptive this method name is. You have other methods with which you use to parse other inputs. 

Perhaps consider renaming this method to ```parseCommand``` instead?

Nephelite commented on src/main/java/pluto/TaskList.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    public int nTasks() {
```
I didn't like this method name, because it is a public method with an undescriptive name. 

Perhaps a more descriptive method name would be better.

Nephelite commented on src/main/java/pluto/Ui.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Ui {
+    private static final String CHATBOT = "pluto.Pluto";
+    private Scanner sc;
```
I didn't like this Scanner object declaration's placement. Perhaps consider moving it below or above all the constants? It's not a coding style violation, I just have OCD.

pratham31012002 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Nephelite commented on src/main/java/pluto/command/DeleteCommand.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws PlutoException {
+        Task t = tasks.deleteTask(idx);
+        try {
+            storage.rewriteFile(tasks);
+            StringBuilder deleteMessage = new StringBuilder();
+            deleteMessage.append("\tNoted. I've removed this task:\n");
+            deleteMessage.append(String.format("\t\t%s\n", t.toString()));
+            deleteMessage.append(String.format("\tNow you have %d tasks in the list.", tasks.nTasks()));
+            ui.print(deleteMessage);
+        } catch (IOException e) {
+            tasks.addTask(idx - 1, t);
+            throw new PlutoException("\tOOPS!!! Couldn't delete task. Try again!");
```
I didn't like this solution for deleting a task from TaskList. 

Consider the scenario where your try block throws an Exception that is not IOException. Perhaps it throws an unchecked exception. In this scenario, you would lose the task to be added back into the TaskList because the catch block is not executed. This is a prime example of using Exceptions as control flow, which is something you should avoid.

Perhaps consider separating the file rewriting process from the task deletion process. The TaskList failing to be stored or rewritten should not mean that the delete command fails entirely.

Nephelite commented on src/main/java/pluto/command/UpdateStatusCommand.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws PlutoException {
+        tasks.markTask(idx, isDone);
+        try {
+            storage.rewriteFile(tasks);
+            StringBuilder markMessage = new StringBuilder();
+            if (isDone) {
+                markMessage.append("\tNice! I've marked this task as done:\n");
+            } else {
+                markMessage.append("\tOK, I've marked this task as not done yet:\n");
+            }
+            markMessage.append("\t\t" + tasks.getTask(idx).toString());
+            ui.print(markMessage);
+        } catch (IOException e) {
+            tasks.markTask(idx, !isDone);
+            throw new PlutoException("\tOOPS!!! Couldn't update task status. Try again!");
+        }
+    }
```
I like that you combined both Mark and Unmark command into a single command. It is a very efficient use of code. Really clean.

Unfortunately, this code block still shares the issues I had mentioned in DeleteCommand. Refer to my comments there for more details.

Nephelite commented on src/main/java/pluto/task/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    private LocalDateTime by;
```
I feel mixed about the name of this variable. It is technically enough, but I still feel like it should be more descriptive. Could you consider making it more descriptive?

(This also applies to the similar private variable in Event)

Nephelite commented on src/main/java/pluto/task/Task.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+
+    public abstract String toFile();
```
I really like this method that you used to turn tasks into storage. If given the time, I would want to try to do the same as well. The idea is really solid.

My only concern is an ignorant or malicious user. What if they run a command like 
todo try | have to complete project

Nephelite reviewed (changes requested):
Overall I think your project is really clean. and has a lot of great ideas and implementations of OOP concepts. I really liked the combination of MarkCommand and UnmarkCommand into a single UpdateStatusCommand.

However, it makes a few errors in using exceptions for control flow and is also slightly too idealistic regarding the user giving bad inputs. 

Also, there are a few module coding convention violations with wildcard imports and the switch case clauses.

pratham31012002 committed:
Set up Gradle

pratham31012002 committed:
Add JUnit Tests

pratham31012002 commented:
> Overall I think your project is really clean. and has a lot of great ideas and implementations of OOP concepts. I really liked the combination of MarkCommand and UnmarkCommand into a single UpdateStatusCommand.
> 
> However, it makes a few errors in using exceptions for control flow and is also slightly too idealistic regarding the user giving bad inputs.
> 
> Also, there are a few module coding convention violations with wildcard imports and the switch case clauses.

Thank you for your deep insights and valuable feedback. I will fix these issues!

pratham31012002 committed:
Create JAR File

pratham31012002 committed:
Add comments

pratham31012002 committed:
Follow the coding standard

pratham31012002 committed:
Add find command

pratham31012002 committed:
Merge branch 'branch-A-JavaDoc'

pratham31012002 committed:
Merge branch 'branch-A-CodingStandard'

pratham31012002 committed:
Merge branch 'branch-Level-9'

pratham31012002 committed:
Check style

pratham31012002 committed:
Add Launcher for GUI

pratham31012002 committed:
Set Layout for GUI

pratham31012002 committed:
Complete basic GUI

pratham31012002 committed:
Complete GUI

pratham31012002 committed:
Merge branch 'branch-Level-10'

pratham31012002 committed:
Merge branch 'branch-Level-10'

pratham31012002 committed:
Add assertions

Classes Parser and Storage access indices of an array at
several places but there's no check to ensure that those
indices can be accessed.

Accessing invalid indices might cause the code to break.

Adding assertions allows to ensure invalid indices
aren't being accessed during testing.

pratham31012002 committed:
Fix JavaDoc

JavaDoc header messages don't follow the standard in certain
places.

JavaDoc comments should be uniform for better code readability.

Let's change the comments to comply with the standards.

pratham31012002 committed:
Improve Code Quality

All display messages are written in the execute() methods of
each individual command.

Adding all these display messages to the Ui class can help in
modifying these messages in the future and keep the display
messages separate from the logic of the code.

Let's add all display messages to the Ui class to keep them
separate from the actual logic of the code.

pratham31012002 committed:
Use Streams

Filter() method of TaskList is too long and nested.

This reduces code readability.

Let's use streams to fix this issue.

pratham31012002 committed:
Merge pull request #2 from pratham31012002/branch-A-Assertions

Branch a assertions

pratham31012002 committed:
Merge branch 'master' of https://github.com/pratham31012002/ip

pratham31012002 committed:
Merge branch 'master'

pratham31012002 committed:
Merge branch 'master'

pratham31012002 committed:
Merge pull request #3 from pratham31012002/branch-A-CodeQuality

Improve Code Quality

pratham31012002 committed:
Merge branch 'master' into branch-A-Streams

pratham31012002 committed:
Merge pull request #4 from pratham31012002/branch-A-Streams

Branch a streams

pratham31012002 committed:
Improve code quality and add more assertions

pratham31012002 committed:
Set up CI

pratham31012002 committed:
Fix testcases

pratham31012002 committed:
Add more JavaDocs

pratham31012002 committed:
Add reschedule task

pratham31012002 committed:
Merge branch 'branch-B-RecurringTasks'

pratham31012002 committed:
Add help command

pratham31012002 committed:
Improve GUI

pratham31012002 committed:
Fix JavaDoc

pratham31012002 committed:
Give a unique personality

pratham31012002 committed:
Handle more errors

pratham31012002 committed:
Add more tests

pratham31012002 committed:
Add Title

pratham31012002 committed:
Add Ui.png

pratham31012002 committed:
Implement C-BetterSearch

pratham31012002 committed:
Implement C-Help

pratham31012002 committed:
Add data folder

pratham31012002 committed:
Improve help Ui

pratham31012002 committed:
Update screenshot

pratham31012002 committed:
Add User Guide

pratham31012002 committed:
Update User Guide

pratham31012002 committed:
Update User Guide

pratham31012002 committed:
Fix rendering issue on Windows

pratham31012002 committed:
Fix issues during initialization

pratham31012002 committed:
Update User Guide and README.md

pratham31012002 committed:
Add finishing touches
